pkg/util/startup/retry.go

```diff
@@ -0,0 +1,205 @@
+// retry should be performed explicitly while using WithoutChecks context
+// to suppress safety mechanisms.
+func RunIdempotentWithRetry(
+	ctx context.Context, opName string, f func(ctx context.Context) error,
+) error {
+	_, err := RunIdempotentWithRetryEx(ctx, opName, func(ctx context.Context) (any, error) {
+		return nil, f(ctx)
+	})
+	return err
+}
+
+// RunIdempotentWithRetryEx run function returning value with startup retry.
+// See RunIdempotentWithRetry for important details.
+func RunIdempotentWithRetryEx[T any](
+	ctx context.Context, opName string, f func(ctx context.Context) (T, error),
+) (T, error) {
+	ctx = context.WithValue(ctx, startupRetryKey{}, "in retry")
+	every := log.Every(5 * time.Second)
+	// Retry failures indefinitely until context is cancelled.
+	var result T
+	var err error
+	for r := retry.StartWithCtx(ctx, startupRetryOpts); r.Next(); {
+		result, err = f(ctx)
+		if err == nil {
+			break
+		}
+		if ctx.Err() != nil || !IsRetryableReplicaError(err) {
+			break
+		}
+		if every.ShouldLog() {
+			log.Infof(ctx, "failed %s during node startup, retrying %s", opName, err)
+		}
+	}
+	return result, err
+}
+
+// Begin ensures that startup detection is correctly identifying startup stack
+// and initializing internal assertions for tests. It must be called first
+// on server startup before any calls to kv or sql are made.
+// Returned function must be called when startup function is complete.
+func Begin(ctx context.Context) func() {
+	if !util.RaceEnabled {
+		return func() {}
+	}
+
+	// Ensure we don't try to call add twice by mistake before inserting our ID.
+	startupID := goid.Get()
+	startupGoroutineIDs.Lock()
+	defer startupGoroutineIDs.Unlock()
+	if _, ok := startupGoroutineIDs.ids[startupID]; ok {
+		log.Fatal(ctx, "startup.Begin() is called twice")
+	}
+	startupGoroutineIDs.ids[startupID] = true
+
+	// Maintain counter as atomic to avoid doing full locks on every send
+	// operation.
+	runningStartup.Add(1)
+	return func() {
+		startupGoroutineIDs.Lock()
+		delete(startupGoroutineIDs.ids, startupID)
+		startupGoroutineIDs.Unlock()
+		runningStartup.Add(-1)
+	}
+}
+
+// AssertStartupRetry is called by dist sender and internal sql executor to
+// ensure that server startup is always performing retry of kv operations.
+func AssertStartupRetry(ctx context.Context) {
+	if !util.RaceEnabled {
+		return
+	}
+	if runningStartup.Load() < 1 {
+		return
+	}
+
+	rv := ctx.Value(startupRetryKey{})
+	if rv == nil && inStartup() {
+		log.Fatal(ctx, "startup query called outside of startup retry loop. See util/startup/retry.go docs")
+	}
+}
+
+// inStartup returns true if currently running go routine called Begin before.
+func inStartup() bool {
+	currentID := goid.Get()
+	startupGoroutineIDs.RLock()
+	defer startupGoroutineIDs.RUnlock()
+	return startupGoroutineIDs.ids[currentID]
+}
+
+// IsRetryableReplicaError returns true for replica availability errors.
+func IsRetryableReplicaError(err error) bool {
```

Comment:
How does this work with SQL queries at start-up? Are we getting the `*kvpb.Error` back in all cases instead of a mangled/flattened SQL error?